parlai/agents/transformer/image_polyencoder.py

```diff
@@ -0,0 +1,184 @@
+from parlai.core.torch_agent import Batch
+from parlai.core.torch_image_agent import TorchImageAgent
+
+
+class ImagePolyencoderAgent(PolyencoderAgent, TorchImageAgent):
+    """
+    Poly-encoder Agent that ingests image features.
+
+    Agent that allows encoding image features and adding or concatenating them to the
+    context encoding.
+    """
+
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        """
+        Add command-line arguments specifically for this agent.
+        """
+        PolyencoderAgent.add_cmdline_args(argparser)
+        TorchImageAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('Image Encoder Args')
+        agent.add_argument(
+            '--image-combination-mode',
+            type=str,
+            default='prepend',
+            choices=['add', 'append', 'prepend'],
+            help='How to combine image embedding (if used) with context embedding',
+        )
+        agent.set_defaults(reduction_type=None)
+        # This agent doesn't support any encoder output reductions
+        return agent
+
+    def build_model(self, states=None):
+        """
+        Return built model.
+        """
+        return ImagePolyencoderModule(self.opt, self.dict, self.NULL_IDX)
+
+    def batchify_image_features(self, batch: Batch) -> Batch:
+        """
+        Return the image features as a Tensor of the correct type.
+
+        Fill in missing feature vectors.
+        """
+
+        # Checks/formatting of batch.image
+        bsz = batch.text_vec.size(0)
+        if batch.image is None or len(batch.image) == 0:
+            batch.image = [None] * bsz
+        else:
+            assert len(batch.image) == bsz
+
+        # Process all image feature vectors, or add in zero vectors if missing
+        processed_features_list = []
+        processed_zero_features = self._process_image_features(
+            torch.zeros((self.image_features_dim,))
+        )
+        for orig_features in batch.image:
+            if orig_features is None:
+                processed_features_list.append(processed_zero_features)
+            elif isinstance(orig_features, torch.Tensor):
+                processed_features_list.append(
+                    self._process_image_features(orig_features)
+                )
+            else:
+                raise ValueError('Unsupported image feature format!')
+
+        # Turn into batchsize x image_features_dim for DataParallel
+        batch.image = torch.stack(processed_features_list)
+
+        return batch
+
+    def _model_context_input(self, batch) -> Dict[str, Any]:
+        """
+        Override PolyencoderAgent's context inputs into the model.
+        """
+        return {'ctxt_tokens': batch.text_vec, 'ctxt_image': batch.image}
+
+    def load_state_dict(self, state_dict):
+        """
+        Override to account for weights used for image features.
+        """
+        for tensor in ['dummy_image_enc', 'ones_mask']:
```

Comment:
is this specifically for init model? I wonder if we need an argument for load_state_dict.

Reply:
i believe it is

Reply:
Yeah, this is for `TorchRankerAgent.load(init_model)`. @stephenroller  - what argument (other than `stated_dict`) did you have in mind for this?